### PR TITLE
Wait until the compensation handlers are completed

### LINKED ISSUE #15066: Wait until the compensation handlers are completed
## Description

When a process instance enters a compensation intermediate throw or end event, we invoke the compensation handlers of completed activities. We should wait at the compensation throw event until all invoked compensation handlers are completed. On completing the last compensation handler, we should complete the compensation throw event and continue the process instance.

We may use the compensation subscription to:
- find the compensation throw event of the compensation handler
- check if other compensation subscriptions of the compensation throw event are active

Screenshot from 2023-11-07 10-54-58

requires #14970

### PR DESCRIPTION
**Pull request by nicpuppa**
## Description

When a process instance enters a compensation intermediate throw or end event, we invoke the compensation handlers of completed activities. On completing the last compensation handler, we should complete the compensation throw event and continue the process instance.

## Related issues

closes #15066

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/compensation/CompensationEventExecutionTest.java [hunk 1]
```diff
@@ -261,11 +280,290 @@ public void shouldActivateCompensationHandlerForEndEvent() {
...
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limitToProcessInstanceCompleted())
+        .extracting(
+            r -> r.getValue().getBpmnElementType(),
+            r -> r.getValue().getBpmnEventType(),
+            Record::getIntent,
+            r -> r.getValue().getElementId())
+        .containsSubsequence(
+            tuple(
```
Comment (nicpuppa) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/compensation/CompensationEventExecutionTest.java:415]: Hey @saig0 there are some test failures related to these assertions, I wasn't able to find the issue. I'm still on it

Comment (nicpuppa) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/compensation/CompensationEventExecutionTest.java:415]: ✅

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/compensation/CompensationEventExecutionTest.java [hunk 2]
```diff
@@ -261,17 +280,301 @@ public void shouldActivateCompensationHandlerForEndEvent() {
...
+                BpmnElementType.USER_TASK,
+                BpmnEventType.COMPENSATION,
+                ProcessInstanceIntent.ELEMENT_COMPLETED,
+                "CompensationHandler2"),
+            tuple(
+                BpmnElementType.INTERMEDIATE_THROW_EVENT,
+                BpmnEventType.COMPENSATION,
+                ProcessInstanceIntent.ELEMENT_COMPLETED,
+                "CompensationThrowEvent"));
```
Comment (saig0) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/compensation/CompensationEventExecutionTest.java:514]: :wrench: This assertion is too long. :laughing: 

Let's reduce it to the major parts. The other aspects are covered by other test cases.

```suggestion
        .containsSubsequence(
            tuple(
                BpmnElementType.INTERMEDIATE_THROW_EVENT,
                BpmnEventType.COMPENSATION,
                ProcessInstanceIntent.ELEMENT_ACTIVATED,
                "CompensationThrowEvent"),
            tuple(
                BpmnElementType.USER_TASK,
                BpmnEventType.COMPENSATION,
                ProcessInstanceIntent.ELEMENT_COMPLETED,
                "CompensationHandler"),
            tuple(
                BpmnElementType.USER_TASK,
                BpmnEventType.COMPENSATION,
                ProcessInstanceIntent.ELEMENT_COMPLETED,
                "CompensationHandler2"),
            tuple(
                BpmnElementType.INTERMEDIATE_THROW_EVENT,
                BpmnEventType.COMPENSATION,
                ProcessInstanceIntent.ELEMENT_COMPLETED,
                "CompensationThrowEvent"),
            tuple(
                BpmnElementType.PROCESS,
                BpmnEventType.UNSPECIFIED,
                ProcessInstanceIntent.ELEMENT_COMPLETED,
                PROCESS_ID));
```

**Review by saig0 — CHANGES_REQUESTED** [review]
@nicpuppa nice job. :rocket: 

I have a few minor comments but nothing major. :+1: 

:x: The commit-lint check failed.
:x: The test results reported one failing test. However, I don't see which test failed. 

Please have a look. :cookie:

**Comment by nicpuppa** [comment]
> ❌ The commit-lint check failed. 

I'm using a new commit plugin for intellJ, I still need to figure out how it works 🤣 I will fix it 👍

File: engine/src/main/java/io/camunda/zeebe/engine/state/mutable/MutableCompensationSubscriptionState.java [hunk 3]
```diff
@@ -15,4 +15,6 @@ public interface MutableCompensationSubscriptionState extends CompensationSubscr
   void put(final long key, CompensationSubscriptionRecord compensation);
 
   void update(final long key, CompensationSubscriptionRecord compensation);
+
+  void remove(String tenantId, long processInstanceKey, String compensableActivityId);
```
[the remark this answers is not in the document]
Comment (nicpuppa) [reply engine/src/main/java/io/camunda/zeebe/engine/state/mutable/MutableCompensationSubscriptionState.java:19]: > Using the subscription key is also safer (for later) to ensure that we delete only one subscription. An activity with a compensation boundary event can be completed multiple times, for example, a multi-instance activity. When the compensation handler is completed, we want to remove only the associated subscription.

Can you clarify this suggestion ? The method for deleting the record from the column family accept only the composite key.

**Comment by nicpuppa** [comment]
Hey @saig0 I applied your suggestion, there is one open point that I would like to discuss. Please have another look 👀

File: engine/src/main/java/io/camunda/zeebe/engine/state/mutable/MutableCompensationSubscriptionState.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (saig0) [reply engine/src/main/java/io/camunda/zeebe/engine/state/mutable/MutableCompensationSubscriptionState.java]: Yes. We could ignore this comment for now. :sweat_smile: 

Currently, we store the subscriptions by `<tenant-id, process-instance-key, activity-id>`. So, we don't have more than one subscription per activity. So far so good. :+1: 

Later, when we support multi-instance, we need to change the state to store multiple subscriptions for an activity. At this point, we will need the key to delete the subscription.

File: engine/src/main/java/io/camunda/zeebe/engine/state/compensation/DbCompensationSubscriptionState.java [hunk 4]
```diff
@@ -75,6 +78,44 @@ public Set<CompensationSubscription> findSubscriptionsByProcessInstanceKey(
...
+        ((key, value) -> {
+          if (value.getRecord().getCompensationHandlerId().equals(compensationHandlerId)) {
+            compensationSubscription.add(value.copy());
+          }
+        }));
+
+    if (compensationSubscription.isEmpty()) {
+      return Optional.empty();
+    }
+    return compensationSubscription.stream().findFirst();
```
Comment (saig0) [inline engine/src/main/java/io/camunda/zeebe/engine/state/compensation/DbCompensationSubscriptionState.java:99]: :wrench: This check is not necessary.

```suggestion
    return compensationSubscription.stream().findFirst();
```

File: engine/src/main/java/io/camunda/zeebe/engine/state/immutable/CompensationSubscriptionState.java [hunk 5]
```diff
@@ -17,4 +18,10 @@ CompensationSubscription get(
 
   Set<CompensationSubscription> findSubscriptionsByProcessInstanceKey(
       String tenantId, long processInstanceKey);
+
+  Optional<CompensationSubscription> findSubscriptionByCompensationHandlerId(
+      String tenantId, long processInstanceKey, String compensationHandlerId);
+
+  Set<CompensationSubscription> findSubscriptionByThrowEventInstanceKey(
```
Comment (saig0) [inline engine/src/main/java/io/camunda/zeebe/engine/state/immutable/CompensationSubscriptionState.java:25]: :wrench: The method returns more than one subscription.

```suggestion
  Set<CompensationSubscription> findSubscriptionsByThrowEventInstanceKey(
```

File: protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/compensation/CompensationSubscriptionRecord.java [hunk 6]
```diff
@@ -145,6 +154,12 @@ public CompensationSubscriptionRecord setCompensableActivityId(
     return this;
   }
 
+  public CompensationSubscriptionRecord setCompensationActivityElementId(
+      final String compensationActivityElementId) {
+    compensationActivityElementIdProperty.setValue(compensationActivityElementId);
+    return this;
+  }
```
Comment (saig0) [inline protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/compensation/CompensationSubscriptionRecord.java:161]: :wrench: We should align the name of the setter, the property, and the property name to `compensationHandlerId`.

Only the getter has been renamed so far. :smile:

**Review by saig0 — APPROVED** [review]
@nicpuppa nice. :tada: 

Just a few minor comments. Please have a look before merging the PR. :rocket:

File: protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/compensation/CompensationSubscriptionRecord.java (hunk 6 above)
Comment (nicpuppa) [reply protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/compensation/CompensationSubscriptionRecord.java:161]: You are right 🤣 🙇

**Comment by nicpuppa** [comment]
bors merge

**Comment by nicpuppa** [comment]
bors retry

**Comment by nicpuppa** [comment]
bors retry

**Comment by nicpuppa** [comment]
bors retry

**Pull request merged**